### ISSUE TITLE
Get prediction from tflite model Android

### ISSUE THREAD

**Issue by abdou31** [tensorflow/tensorflow#30410]
I worked on eye detection in real time project.
I have generated a tflite model graph from trained frozen inference graph .pb and I want to use the tflite model in Android Studio.
I have integrated OpenCV library and I have added TensorFlow tflite library in the application.
The front cam of my android device works fine.
+ I have this code: 
```.java

package com.example.opencvtest;

import org.opencv.android.BaseLoaderCallback;
import org.opencv.android.CameraBridgeViewBase.CvCameraViewFrame;
import org.opencv.android.LoaderCallbackInterface;
import org.opencv.android.OpenCVLoader;
import org.opencv.core.Mat;
import org.opencv.android.CameraBridgeViewBase;
import org.opencv.android.CameraBridgeViewBase.CvCameraViewListener2;

import android.app.Activity;
import android.content.pm.ActivityInfo;
import android.content.res.AssetFileDescriptor;
import android.hardware.Camera;
import android.os.Bundle;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.SurfaceView;
import android.view.WindowManager;
import android.widget.Toast;

import java.io.FileInputStream;
import java.io.IOException;
import java.lang.reflect.Method;
import java.nio.MappedByteBuffer;
import java.nio.channels.FileChannel;
import org.tensorflow.lite.Interpreter;
import org.tensorflow.lite.Tensor;


public class MainActivity extends Activity implements CvCameraViewListener2 {
    private static final String TAG = "OCVSample::Activity";

    private CameraBridgeViewBase mOpenCvCameraView;
    private boolean              mIsJavaCamera = true;
    private MenuItem             mItemSwitchCamera = null;
    String modelFile="iristf.tflite";
    Interpreter tflite;
    Tensor ts;


    //Load model tflite
    private MappedByteBuffer loadModelFile(Activity activity, String MODEL_FILE) throws IOException {
        AssetFileDescriptor fileDescriptor = activity.getAssets().openFd(MODEL_FILE);
        FileInputStream inputStream = new FileInputStream(fileDescriptor.getFileDescriptor());
        FileChannel fileChannel = inputStream.getChannel();
        long startOffset = fileDescriptor.getStartOffset();
        long declaredLength = fileDescriptor.getDeclaredLength();
        return fileChannel.map(FileChannel.MapMode.READ_ONLY, startOffset, declaredLength);
    }
    private BaseLoaderCallback mLoaderCallback = new BaseLoaderCallback(this) {
        @Override
        public void onManagerConnected(int status) {
            switch (status) {
                case LoaderCallbackInterface.SUCCESS:
                {
                    Log.i(TAG, "OpenCV loaded successfully");
                    mOpenCvCameraView.setCameraIndex(1);
                    mOpenCvCameraView.enableView();
                } break;
                default:
                {
                    super.onManagerConnected(status);
                } break;
            }
        }
    };

    public MainActivity() {
        Log.i(TAG, "Instantiated new " + this.getClass());
    }

    /** Called when the activity is first created. */
    @Override
    public void onCreate(Bundle savedInstanceState) {
        Log.i(TAG, "called onCreate");
        super.onCreate(savedInstanceState);
        setRequestedOrientation(ActivityInfo.SCREEN_ORIENTATION_PORTRAIT);

        getWindow().addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON);

        setContentView(R.layout.activity_main);
        try {
            tflite = new Interpreter(loadModelFile(this, modelFile));
             ts=tflite.getInputTensor(0);

        }
        catch (IOException e) {
            e.printStackTrace();
        }

        mOpenCvCameraView = (CameraBridgeViewBase) findViewById(R.id.tutorial1_activity_java_surface_view);

        mOpenCvCameraView.setVisibility(SurfaceView.VISIBLE);

        mOpenCvCameraView.setCvCameraViewListener(this);







    }

    @Override
    public void onPause()
    {
        super.onPause();
        if (mOpenCvCameraView != null)
            mOpenCvCameraView.disableView();
    }

    @Override
    public void onResume()
    {
        super.onResume();
        if (!OpenCVLoader.initDebug()) {
            Log.d(TAG, "Internal OpenCV library not found. Using OpenCV Manager for initialization");
            OpenCVLoader.initAsync(OpenCVLoader.OPENCV_VERSION_3_0_0, this, mLoaderCallback);
        } else {
            Log.d(TAG, "OpenCV library found inside package. Using it!");
            mLoaderCallback.onManagerConnected(LoaderCallbackInterface.SUCCESS);
        }
    }

    public void onDestroy() {
        super.onDestroy();
        if (mOpenCvCameraView != null)
            mOpenCvCameraView.disableView();
    }

    public void onCameraViewStarted(int width, int height) {
    }

    public void onCameraViewStopped() {
    }

    public Mat onCameraFrame(CvCameraViewFrame inputFrame) {
        return inputFrame.rgba();
    }
    protected void setDisplayOrientation(Camera camera, int angle){
        Method downPolymorphic;
        try
        {
            downPolymorphic = camera.getClass().getMethod("setDisplayOrientation", new Class[] { int.class });
            if (downPolymorphic != null)
                downPolymorphic.invoke(camera, new Object[] { angle });
        }
        catch (Exception e1)
        {
            e1.printStackTrace();
        }
    }
}


```

My goals are : 

+ Decomposed the input streaming from cam into frames.
+ Get predictions from frames using tflite model.

I have followed many tutorials ( Medium StackOverflow, Github, Tensorflow Documentation...).
But I didn't get an answer to those questions below :

I want to know how can I get predictions from video stream?
How can I get frames from input streaming?
How to use the tflite model to get predictions from the video stream ( or images )?

Thanks.

**Comment by rmothukuru** [comment]
This question is better asked on [StackOverflow](http://stackoverflow.com/questions/tagged//tensorflow) since it is not a bug or feature request. There is also a larger community that reads questions there.

**Comment by abdou31** [comment]
I ask this  question on stackoverflow but I didn't get any answer

**Comment by rmothukuru** [comment]
@abdou31 ,
Sorry to inform you that providing support at such level will be difficult as there are high priority Bugs, Performance Issues and Feature Requests to be looked upon.

**Comment by abdou31** [comment]
But this is related to developer's use , I mean my question have a relationship with how to use Tensorflow on mobile ( documentation ) 
But I didn't found anything on documentation.
The only thing that I found, is ML kit with Tensorflow ( firebase ) but this is not recomended beacause it not always free and it requires INTERNET.

**Comment by anilmaddala** [comment]
@abdou31 Which model are you training? What is the input data to the model?

From my understanding, there is no way to run the model on Videos (Not sure if there are models which work on videos). The models usually expect images or sequence of images. 

So you have to extract the images from the video feed and input to your inference pipeline

**Comment by abdou31** [comment]
The model that I worked on is Resnet,
This is the project that I use for creating my own model and training my own dataset 
+ [link](https://github.com/yinguobing/cnn-facial-landmark)
The input data is a sequence of images ( or streaming input e.g webcam ).
The output sequence of image ( or streaming input e.g webcam that show the points on the eye region ) like this :
![ssasse](https://user-images.githubusercontent.com/19480228/60874720-c0eaea80-a238-11e9-9b81-3de29a965879.PNG)

I have successfully got a result from my webcam ( pc webcam ) using frozen inference model ( not good result because the training is started from scratch).
But the problem now is on the mobile device, I need to use the model ( tflite or frozen .pb ) with Android SDK.
How can I do that?

**Comment by anilmaddala** [comment]
Checkout how TFLite andorid sample apps. You basically get imae from camera and do the required preprocessing before passing the image to your model.

**Comment by abdou31** [comment]
I started to do this
I have asked for help from others on many websites (StackOverflow, data StackExchange ) but I didn't get any results 
As you can see, those are links related to my questions on websites I talked about:
[link 1](https://datascience.stackexchange.com/questions/55089/extract-features-predictions-from-tflite-model-on-android-device)
[link 2](https://stackoverflow.com/questions/56915664/extracting-features-predictions-from-tflite-model)

On those websites, I have attached the source code that I have got from some tutorials and TensorFlow samples to starting the project. 
The principal problem here is that TensorFlow samples are talking about image classification and object detection but my project is talking about eye region landmarks ( draw landmark on the eye region ) using the regression, not a classification.
Do you have any idea about how can I get predictions from the last layer of the tflite model?

**Comment by anilmaddala** [comment]
Another example you might want to look at in Image segmentation on Andorid, there are different projects on github which do this.

**Comment by abdou31** [comment]
Also image segmentation is not related to my project..
Can you gives me links of project on github that can help me?

**Comment by jdduke** [comment]
Hi @abdou31 , the detection sample is not unlike what you want (see https://www.tensorflow.org/lite/models/object_detection/overview), and demonstrates how to feed frame inputs and parse semantic outputs (e.g., labels, boxes) for use in your app.

**Comment by abdou31** [comment]
I had explained what I want and I had told you that the project is not a classification [link](https://github.com/tensorflow/examples/pull/73) my CNN used regression to predict and localize the eye region and draw landmarks on contours of the eye.
So image classification doesn't meet what I want.

**Comment by jdduke** [comment]
@abdou31 while classification/detection are not identical to your use-case, the point is that they demonstrate how to map to and from model-specific inputs/outputs. I would recommend using some of the instructions @ https://www.tensorflow.org/lite/guide/faq#how_do_i_inspect_a_tflite_file, or a model visualizer like [Netron](https://github.com/lutzroeder/netron), to inspect the names/types of the inputs/outputs of your graph, and work back from there.

**Comment by abdou31** [comment]
Thanks, I have the name and the types of the outputs and inputs of my model ( frozen inference graph .pb ) that I used to generate tflite model.
I have read and I have seen many tutorials and I know that my tflite model has an input tensor with those parameters (batch, width, height, channels) => (1,112,112,3): input_to_float:0
channels=>3 ( RGB) .
and for the input is a frame ( image ):  logits/BiasAdd:0
I had successfully got a good result and extract predictions from frozen inference graph using a python script.
When I take the webcam as an input, I get an output with landmark drawn on the eye region.
So my object, for now, not a detection from webcam of the PC, but from my smartphone.
For that reason, I have converted the model to tflite model to use it in Android Studio following steps and instructions mentioned on the Tensorflow guide ( add the model in the assets folder...).
But I encountered some difficulties when I try to add an image as an input and try to get just the positions of landmarks ( x and y coordinates ).
Hoping you understand me.

**Comment by yurnery** [comment]
@abdou31 I got the same problem as you. In my case, I want to extract the image features before the fully connected layer by TFLite, and I will use the image feature to find some similar images. But the solutions what I found is all about Python and C++, I don't know how to get the feature by TFLite, did you solve it?

**Comment by abdou31** [comment]
I change to CoreML with iPhone. 
I did not find a solution for tensorflow lite

**Comment by yurnery** [comment]
@abdou31 So your project dose not support Android right now? Or you use other ML framework on Android? And could you tell me which API in CoreML can extract the features, thanks.

**Comment by abdou31** [comment]
I'm a Phd student and the project is to validate the end of the studies.
So the project requires a mobile app ( for iphone or for android ).
For that reason I have changed to Iphone ( CoreML), also I have found a implementation that does not needed a many changes.
